client/driver/executor/executor.go

```diff
@@ -618,3 +673,83 @@ func (e *UniversalExecutor) interpolateServices(task *structs.Task) {
 		service.Tags = e.ctx.TaskEnv.ParseAndReplace(service.Tags)
 	}
 }
+
+// collectPids collects the pids of the child processes that the executor is
+// running every 5 seconds
+func (e *UniversalExecutor) collectPids() {
+	for {
+		timer := time.NewTimer(pidScanInterval)
+		select {
+		case <-timer.C:
+			pids, err := e.getAllPids()
+			timer.Reset(pidScanInterval)
+			if err != nil {
+				e.logger.Printf("[DEBUG] executor: error collecting pids: %v", err)
+			}
+			e.pids = pids
+		case <-e.processExited:
+			timer.Stop()
+			return
+		}
+	}
+}
+
+// scanPids scans all the pids on the machine running the current executor and
+// returns the child processes of the executor.
+func (e *UniversalExecutor) scanPids() ([]*NomadPid, error) {
```

Comment:
This function isn't quite right. It should be using DFS algorithm. We miss double forks.

Parent -> Child -> Grand Child -> ...